notebooks/notebook_template_review.py

```diff
@@ -595,51 +504,280 @@ def parse_recommendations(path: str,
     return nth - 1
 
 
-def get_cell(path: str, 
-             cells: list, 
-             nth: int) -> (list, int):
+def parse_dataset(path: str, 
+                  cells: list, 
+                  nth: int) -> int:
+    
     """
-        Get the next notebook cell.
-        
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
-        nth: The index of the last cell that was parsed (reviewed).
-        
-        Returns:
-        
-        cell: content of the next cell
-        nth + 1: index of subsequent cell
+    Parse the dataset cell
+    
+    path: used only for reporting an error
+    cells: The content cells (JSON) for the notebook
+    nth: The index of the last cell that was parsed (reviewed).
+    
+    Returns: cell index
     """
-    while empty_cell(path, cells, nth):
-        nth += 1
+    # Dataset
+    cell, nth = get_cell(path, cells, nth)
+    if not cell['source'][0].startswith("### Dataset") and not cell['source'][0].startswith("### Model") and not cell['source'][0].startswith("### Embedding"):
+        report_error(path, ErrorCodes.ERROR_DATASET_NOTFOUND, "Dataset/Model section not found")
         
-    cell = cells[nth]
-    if cell['cell_type'] == 'markdown':
-        check_text_cell(path, cell)
-    return cell, nth + 1
+    return nth
+
+def parse_costs(path: str, 
```

Comment:
1. Instead of having named functions that are directly called by `parse_notebook`, create an abstractbaseclass `Rule`.

```
class Rule(abc.ABC):
    def validate(self, path: str, cells: list, nth_please_rename: int) -> Optional[ErrorCode]:
        pass
```

2. Have each rule implement that ABC:

```
class CopyrightRule(Rule):
    def validate(self, path: str, cells: list, nth_please_rename: int) -> Optional[ErrorCode]:
        ...
```

3. Move rules to another file.

Reply:
4. Ideally, add tests for each rule.
5. Move notebook_template_review to its own folder.
6. Separate linting code and other requirements (index generation, etc) to separate files.

Reply:
7. Either maintain a list of rules or autodetect Rule classes.

Reply:
8. `parse_notebook` should look something like:

def parse_notebook(...):
    ...
    error_codes = [for rule.validate(...) in RULES]